File: 97. WordLadder.java
import java.util.ArrayDeque;
import java.util.HashSet;
import java.util.Queue;
import java.util.Set;

public class WordLadder {

    public static int ladderLength(String beginWord, String endWord, java.util.List<String> wordList) {
        Set<String> words = new HashSet<>(wordList);
        if (!words.contains(endWord)) {
            return 0;
        }

        Queue<String> queue = new ArrayDeque<>();
        Set<String> visited = new HashSet<>();

        queue.offer(beginWord);
        visited.add(beginWord);

        int length = 1;

        while (!queue.isEmpty()) {
            int size = queue.size();

            for (int i = 0; i < size; i++) {
                String current = queue.poll();
                if (current.equals(endWord)) {
                    return length;
                }

                char[] chars = current.toCharArray();
                for (int j = 0; j < chars.length; j++) {
                    char original = chars[j];

                    for (char ch = 'a'; ch <= 'z'; ch++) {
                        if (ch == original) {
                            continue;
                        }

                        chars[j] = ch;
                        String next = new String(chars);

                        if (words.contains(next) && !visited.contains(next)) {
                            visited.add(next);
                            queue.offer(next);
                        }
                    }

                    chars[j] = original;
                }
            }

            length++;
        }

        return 0;
    }

    public static void main(String[] args) {
        System.out.println(ladderLength("hit", "cog", java.util.List.of("hot", "dot", "dog", "lot", "log", "cog")));
        System.out.println(ladderLength("hit", "cog", java.util.List.of("hot", "dot", "dog", "lot", "log")));
        System.out.println(ladderLength("a", "c", java.util.List.of("a", "b", "c")));
    }
}

File: 97. WordLadder.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <unordered_set>
using namespace std;

int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
    unordered_set<string> words(wordList.begin(), wordList.end());
    if (!words.count(endWord)) return 0;

    queue<string> q;
    unordered_set<string> visited;

    q.push(beginWord);
    visited.insert(beginWord);

    int length = 1;

    while (!q.empty()) {
        int size = q.size();
        for (int i = 0; i < size; i++) {
            string current = q.front();
            q.pop();
            if (current == endWord) return length;
            for (int j = 0; j < (int)current.size(); j++) {
                char original = current[j];
                for (char ch = 'a'; ch <= 'z'; ch++) {
                    if (ch == original) continue;
                    current[j] = ch;
                    if (words.count(current) && !visited.count(current)) {
                        visited.insert(current);
                        q.push(current);
                    }
                    current[j] = original;
                }
            }
        }
        length++;
    }

    return 0;
}

int main() {
    vector<string> wl1 = {"hot","dot","dog","lot","log","cog"};
    cout << ladderLength("hit", "cog", wl1) << "\n";
    vector<string> wl2 = {"hot","dot","dog","lot","log"};
    cout << ladderLength("hit", "cog", wl2) << "\n";
    vector<string> wl3 = {"a","b","c"};
    cout << ladderLength("a", "c", wl3) << "\n";
    return 0;
}
